[2단계 - 웹 기반 로또 게임] 우디(박연기) 미션 제출합니다.

yeongipark opened the pull request:
안녕하세요, 하리! !! ! ! !! !!

벌써 마지막 리뷰라니… 🥲🥲🥲 시간이 정말 빠르네요!

이전까지는 구현하면서 고민한 내용을 따로 메모하지 않고 넘어가다 보니, PR에 작성할 내용이 많지 않았는데, 이번에는 고민한 내용을 기록하면서 구현해서 그런지 이번 PR에서는 제가 어떤 고민을 했는지를 더 잘 정리할 수 있었던 것 같아요!

하리가 제 고민을 읽어보시고 잘못된 부분이나 더 나은 방향이 있다면 피드백 부탁드립니다!

이번 리뷰도 잘 부탁드리겠습니다! 😊😊😊😊

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x]  기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x]  기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x]  테스트 코드는 모두 정상적으로 실행되나요?
- [x]  (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
    - 배포한 링크 기입: https://yeongipark.github.io/javascript-lotto/

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.

이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

# 1) 이번 단계에서 가장 많이 고민했던 문제

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->

## **1️⃣ 웹뷰 관련 코드의 파일 구조 및 폴더 위치**

처음에는 **웹뷰 관련 기능을 모두 step2-index.js 파일에 작성**했습니다. 이후 리팩토링 과정에서, 이벤트별로 파일을 분리하는 것이 필요하다고 판단했습니다.

로또 게임에서 발생하는 주요 이벤트는 다음과 같습니다.

- **구매 관련 이벤트**: 구입 금액 입력 후 구매 버튼 클릭
- **결과 확인 이벤트**: 당첨 번호 및 보너스 번호 입력 후 결과 확인 버튼 클릭
- **재시작 이벤트**: 다시 시작 버튼 클릭
- **모달 관련 이벤트**: 모달 창 닫기

이를 바탕으로 **각 이벤트별 핸들러 파일**을 아래와 같이 분리했습니다.

- buyHandler.js → 구매 이벤트 처리
- resultHandler.js → 결과 확인 이벤트 처리
- modalHandler.js → 모달 창 이벤트 처리
- restartHandler.js → 게임 재시작 이벤트 처리

또한, **HTML 노드를 저장한 변수를 모아 관리할 필요성**을 느껴 dom.js 파일을 만들어 해당 요소들을 정리했습니다.

그리고 LottoMahcine의 인스턴스를 resultHandler.js와 buyHandle.js가 공용으로 동일한 인스턴스를 사용해야 하기 때문에 lottoState라는 객체를 만들어 해당 객체의 프로퍼티로 lottoMachine을 생성해 export하는 방식으로 구현했습니다!

파일을 어느 폴더에 저장할지 고민하다가, **웹뷰와 관련된 코드들이므로 webView 폴더에 저장**했습니다. 하지만 지금 다시 생각해보니, **핸들러 파일들이 주로 이벤트 처리를 담당하므로 eventHandler라는 폴더명이 더 적절할 수도 있겠다는 생각이 드네요 🤔**

##  **2️⃣ HTML 노드를 저장하는 변수를 개별적으로 export할지, 객체로 묶어 export default할지**

처음에는 **각각의 변수를 개별적으로 export**했습니다. 하지만 JavaScript에서 접근해야 하는 변수가 점점 많아지고, 파일을 분리하면서 **import문이 많아지는 문제**가 발생했습니다.

이에 따라 **변수들을 객체로 묶어 export default하는 방식**을 시도했지만, 이 경우 노드를 사용할 때 **객체.변수이름 형태로 점 표기법을 사용해야 해서 가독성이 떨어진다고 느꼈습니다**.

만약 각 파일에서 **import하는 변수가 2~3개 정도였다면 개별적으로 export하는 방식이 더 적절**했을 것 같습니다. 하지만 **일부 파일에서 import 문이 너무 많아지면서 코드의 가독성이 저하된다고 판단**하여, 결국 객체로 묶어서 export하는 방식을 선택했습니다.

이에 대해 **하리의 의견이 궁금합니다!** 😊

## **3️⃣ 어디까지 함수로 분리해야 할까?**

리팩토링하면서 복잡한 로직을 함수로 분리하는 과정중에서 어디까지 함수로 분리해야할까? 에 대해 고민하였습니다. 

```jsx
Array.from(state.paper_winning_number_inputs).map(
      (winningInput) => Number(winningInput.value)
    );
```

예를 들어, 위 코드는 **당첨 번호 입력 필드 6개를 순회하며 값을 숫자 배열로 변환하는 로직**입니다.

처음에는 해당 로직을 **한눈에 이해하기 어렵다고 판단하여 함수로 분리하려고** 했습니다.

하지만 한편으로는 **함수로 분리했을 때와 그대로 두었을 때의 차이가 크지 않을 것** 같다는 생각이 들었습니다.

•	**함수를 찾아가는 시간** vs **로직을 바로 이해하는 시간**이 비슷할 수 있음

•	**재사용되지 않는 함수라면 분리의 이점이 크지 않음**

이러한 이유로 결국 **별도로 분리하지 않고 그대로 두기로 결정**했는데, 이에 대해 **하리의 의견이 궁금합니다!** 😊

##  **4️⃣ HTML 태그를 JavaScript로 렌더링할지 고민**

처음 미션을 접했을 때, **동적으로 렌더링해야 하는 부분만 JavaScript로 처리하는 것이 적절하다고 생각**했습니다. 하지만 **모든 HTML을 JavaScript로 렌더링하는 방식도 고려**했으며, 이 경우 **두 가지 선택지**가 있었습니다.

1.	innerHTML을 사용하여 문자열로 HTML 태그를 삽입

2.	appendChild를 사용하여 DOM 요소를 생성하고 자식 요소로 추가

하지만 각각의 방식에는 단점이 있었습니다.

•	**innerHTML 방식**: 문자열로 태그를 추가하면 **자동완성 기능이 제공되지 않고, HTML 관련 플러그인을 활용하기 어려워 유지보수성이 떨어진다**고 생각했습니다.

•	**appendChild 방식**: 태그를 개별적으로 생성하고 추가하는 방식은 **HTML 구조를 한눈에 파악하기 어려운 단점**이 있었습니다.

반면, **HTML 파일에 직접 태그를 작성하면 구조를 쉽게 파악할 수 있어 유지보수성이 높아진다고 판단**했습니다.

또한, **DOM 조작은 큰 비용이 소모되는 작업이므로 최소화하는 것이 중요하다고 생각하여, 정말 필요한 부분만 JavaScript로 동적으로 렌더링**하기로 결정했습니다.

## **5️⃣ css 컨벤션 지키려고 노력**

이때까지 css도 컨벤션을 지켜서 개발한다는 사실을 모르고 이번에 알았습니다. 그래서 저만의 css 컨벤션 규칙을 만들어서 이번에 적용해보았습니다! 혹시 컨벤션이 제대로 지켜지지 않은 부분은 얘기해 주세요!

**우디의 CSS 컨벤션 속성 작성 순서**

1.	**크기 관련**: width, height

2.	**레이아웃 관련**: display, flex 관련 속성 (flex, justify-content, align-items 등)

3.	**배경 스타일**: background

4.	**여백 관련**: margin, padding

5.	**테두리 관련**: border, border-radius

6.	**텍스트 스타일**: font-size, color, text-align 등

## **6️⃣ CSS import 방식 선택**
CSS를 가져오는 방법으로 두 가지 방식을 고민했습니다.

1. <link> 태그를 사용하여 HTML에서 직접 가져오기
2. CSS 파일 내에서 @import를 사용하여 가져오기

각 방법의 장단점을 비교해본 결과, 1번 방식(link 태그 사용)은 CSS 파일을 비동기적으로 병렬 다운로드하여 페이지 로딩 속도가 더 빠르다는 장점이 있었습니다.
반면, 2번 방식(@import 사용)은 CSS 파일이 순차적으로 로드되기 때문에 성능이 상대적으로 떨어집니다.

이러한 점을 고려하여, 최적의 성능을 위해 1번 방식(<link> 태그 사용)을 선택했습니다.

## **7️⃣ 로또 최대값**
웹에서 View를 구현하면서 구입 금액의 최대값을 조정할 필요가 있었습니다.
로또를 100,000개 이상 구매할 경우 렌더링 속도가 느려지고, 스크롤 시 렉이 발생하는 문제가 있었기 때문입니다.

따라서, 기존에 설정했던 최대값을 조정해야 했고,
실제 로또 게임에서는 1회 구매 한도가 100,000원으로 제한되어 있다는 점을 참고하여,
최대 구매 금액을 100,000원으로 제한하였습니다.

## **8️⃣ 로또 추가 과정에서 발생하는 appendChild 호출 최소화**
```javascript
lottoDiv.appendChild(ticketIcon);
lottoDiv.appendChild(lottoNumbersDiv);
domRefs.$lottoInfo.appendChild(lottoDiv); // 로또를 생성하고 바로 lottoInfo에 추가 (로또 갯수만큼 appendChild 실행)
```
이전의 createLottos 함수에서는 로또를 생성할 때마다 lottoInfo(로또를 담는 컨테이너)에 appendChild를 호출하여, 로또 개수만큼 appendChild가 실행되는 문제가 있었습니다.

하지만 appendChild는 DOM을 직접 조작하는 작업이므로, 호출 횟수를 최소화하는 것이 성능 개선에 중요하다고 판단했습니다.
따라서, 로또를 생성할 때마다 즉시 lottoInfo에 추가하는 것이 아니라, DocumentFragment에 먼저 추가한 후, 모든 로또가 생성된 후에 단 한 번만 appendChild를 실행하도록 변경하여 성능을 개선했습니다.

```javascript
const fragment = document.createDocumentFragment();
lottoDivs.forEach((lottoDiv) => fragment.appendChild(lottoDiv));
domRefs.$lottoInfo.appendChild(fragment); // 한 번만 DOM 조작
```

💡 DocumentFragment를 사용한 이유

DocumentFragment는 가상 DOM 역할을 하는 요소로,
실제 DOM을 조작하는 것이 아니라 메모리에서 관리되기 때문에, 여러 번 appendChild를 호출해도 성능에 영향을 미치지 않습니다.
최종적으로 한 번만 실제 DOM에 추가되므로 렌더링 비용이 줄어들어 성능이 개선됩니다! 

# 2) 이번 리뷰를 통해 논의하고 싶은 부분

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

## **1️⃣ 현업에서 class와 id를 어떻게 구분해서 사용하는지**

저는 평소에 id보다 class를 더 많이 사용합니다!

그 이유는 **처음 리액트와 HTML을 배울 때 인강에서 id를 사용하지 않고 class를 주로 사용했기 때문**입니다.

보통 **id는 고유한 값을 나타낼 때, class는 재사용이 가능할 때 사용한다고 알고 있는데**, 실제로 **현업에서도 이 원칙을 그대로 적용하는지 궁금합니다.**

또한, 제 코드에서는 **한 번만 사용되는 class가 많습니다.**

이 경우 **모두 id로 바꾸는 것이 좋은지**에 대한 하리의 의견이 궁금합니다!

## **2️⃣ css 클래스 이름 짓는법**

저는 보통 **부모 요소의 이름을 앞에 작성한 후, _(언더스코어)로 자식 요소를 구분하는 방식**을 사용합니다.

예를 들어, paper_input처럼 네이밍을 합니다.

이렇게 하는 이유는 **클래스명이 의도치 않게 겹쳐 스타일이 깨지는 것을 방지하기 위해서입니다.**

혹시 **하리는 다른 네이밍 방식을 사용한다면 어떻게 하는지** 궁금하고, **제 방식에 대한 의견도 듣고 싶습니다!**

## **3️⃣ container와 wrap의 차이?**

CSS를 작성하다 보면 container와 wrap을 사용하는 경우가 많습니다.

저는 **최상위 요소를 container로 설정하고, 그 내부를 묶는 용도로 wrap을 사용**합니다.

실제로 **이렇게 사용하는 것이 맞는지**, 혹은 다른 기준이 있는지 궁금합니다!

## **4️⃣ css 컨벤션**

저만의 **CSS 컨벤션을 만들어 처음 적용해 보았는데**, 혹시 **업계에서 정형화된 CSS 컨벤션이 따로 있는지** 궁금합니다!

또한, **하리는 CSS 컨벤션을 어떤 방식으로 적용하는지** 공유해 주시면 감사하겠습니다! 😊

## **5️⃣ step2-index.js은 어떻게 사용하지??**

이번에 Web으로 구현하면서 컨트롤러가 없어지고, addEventListener를 통해 모든 사용자 입력과 출력을 관리하게 되었습니다.
그렇다 보니, 1단계에서와는 달리 step2-index.js 파일을 활용할 방법이 없었습니다.

이전처럼 해당 파일에서 실행을 명시적으로 호출할 필요도 없어, 이번에는 step2-index.js에 아무 내용도 작성하지 않았는데,
하리님께서는 이 파일에 어떤 내용을 넣는 것이 좋을지, 또는 구조를 어떻게 변경하면 좋을지에 대한 의견이 궁금합니다!

## **6️⃣ html 렌더링 방식**
앞서 고민한 내용에서도 언급했듯이, 저는 최대한 HTML에서 직접 구조를 작성하고, 정말 필요한 부분만 JavaScript로 동적으로 렌더링하도록 했습니다.

하지만 많은 크루들이 모든 HTML을 JavaScript로 렌더링하는 방식을 선택하더라고요.
처음에는 제 방식이 더 적절하다고 생각하여 그대로 진행했지만, 점점 주변 크루들이 JavaScript로 렌더링하는 것을 보면서 “내가 틀린 걸까?” 하는 생각이 들었습니다.

그런데 아무리 고민해도 JavaScript로 모든 HTML을 렌더링했을 때의 장점이 명확하게 와닿지 않습니다.
HTML 파일이 깔끔해진다는 점 정도는 이해가 가지만, 그 외에 JavaScript를 통한 렌더링이 가지는 구체적인 이점이 무엇인지 궁금합니다.
또한 해당 방식이 정말 html을 직접 작성하는 것보다 좋은 건지 궁금해요!!

하리는 이 부분에 대해 어떻게 생각하나요 !?!?
---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ]  TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [ ]  도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [ ]  객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [ ]  불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

### 2단계

- [x]  도메인 로직에 불필요한 영향을 주지 않고 UI 변경에 대응했는가?
- [x]  DOM 조작과 이벤트 활용을 JavaScript의 개념에 맞게 이해하고 적절하게 적용했는가?
- [x]  웹 표준을 준수하는 마크업을 활용하며, 스타일 작성에 일관성이 있는가?
</details>

dev-dino22 committed:
docs(README.md): 기능목록 작성

dev-dino22 committed:
feat: 로또 랜덤 생성기 구현

dev-dino22 committed:
feat: 로또 객체 생성(generateLottos() 로또객체 내부로 이동)

dev-dino22 committed:
feat(Winnings): 당첨 번호와 보너스 번호를 검증한 후 저장

dev-dino22 committed:
feat(Winnings): 당첨 통계 계산

dev-dino22 committed:
feat(LottoMachine): 로또머신에 배열 생성 메서드 구현

dev-dino22 committed:
refactor(LottoMachine): getStatistics() 메서드 삭제

dev-dino22 committed:
feat(Validator):메서드  유효성 검사 로직 작성

dev-dino22 committed:
refactor(Winnings): 검증 로직 리팩토링

dev-dino22 committed:
feat: View 작성

dev-dino22 committed:
feat(Error): 에러 메세지 객체 생성

dev-dino22 committed:
refactor: 메세지 상수화

dev-dino22 committed:
feat(Main): 메인 로직 작성

dev-dino22 committed:
refactor: Input, Output 수정

dev-dino22 committed:
feat: Index 작성

dev-dino22 committed:
test: console.log 제거

yeongipark committed:
refactor: validator 리팩토링

- ValidatorUtil을 util로 이동
- 반복 되어 사용되는 로직만 util로 분리
- 각 관심사에 대해 validator를 분리

yeongipark committed:
refactor(validator) : 바뀐 validator 적용

yeongipark committed:
refactor: errorHandler 방식 변경

- 기존의 복잡한 형태의 에러 핸들러 방식을 간단한 방식으로 변경

yeongipark committed:
refactor: 객체 변수명 카멜 케이스로 변경

yeongipark committed:
refactor(main): 다시 시작 부분도 errorHandler 방식으로 변경

yeongipark committed:
refactor: 프로그램의 전체적인 구조 변경

- Main 클래스가 LottoMachine 클래스만을 통해 상호작용할 수 있도록 변경
- Lottos 클래스를 만들어 로또 배열을 저장
- getter를 사용하는 것이 아닌 메시지를 보내 상호작용하도록 구현

yeongipark committed:
refactor: output 메서드 로직 변경

- Main 클래스의 로직을 단순화 하기 위해 ouput 메서드 로직 변경

yeongipark committed:
docs: README.md 파일 설계 추가

- 기능을 세분화 함
- 설계 내용을 추가함

yeongipark committed:
chore: 필요없는 파일 삭제

yeongipark committed:
refactor : 프로그램 상수화

yeongipark committed:
refactor: 응집도를 위해 상수를 사용하는 폴더에 상수 이동

yeongipark committed:
refactor: 큰 숫자 "_" 추가하여 가독성 증가

yeongipark committed:
refactor: 메서드 명 명사형으로 통일

yeongipark committed:
refactor: 덜한 상수화 마저 진행

yeongipark committed:
test(Lotto): Lotto 클래스 테스트 코드 수정

yeongipark committed:
test: 없어진 클래스 테스트 코드 삭제

yeongipark committed:
test: BonusNumberValidator 클래스 테스트 코드 작성

yeongipark committed:
test: Calculator 클래스 테스트 코드 작성

yeongipark committed:
test: Lotto 클래스 테스트 코드 작성

yeongipark committed:
test: NumbersValidator 클래스 테스트 코드 작성

yeongipark committed:
test: PurchasePriceValidator 클래스 테스트 코드 작성

yeongipark committed:
docs: README.md 파일에 설계도 이미지 추가

yeongipark committed:
refactor(Calculator) : reduce를 이용해 로직 단순화

- getStatistics 메서드의 내부 로직에 reduce를 상용해 로직 단순화

yeongipark committed:
chore(PurchasePriceValidator) : 파일명 오타 수정

yeongipark committed:
refactor(PurchasePriceValidator) : 1000 상수화

yeongipark committed:
refactor: Lotto 배열 생성 로직 Lottos로 이동

yeongipark committed:
test: 정상 통과 테스트 로직 추가

yeongipark committed:
chore(view): 파일명 카멜 케이스로 변경

yeongipark committed:
refactor: validator 메서드 명 직관적으로 변경

yeongipark committed:
refactor(Main): 외부에서 사용되지 않는 메서드 프라이빗 메서드로 선언

yeongipark committed:
feat: ui 스파게티 코드 작성

yeongipark committed:
fix(NumbersValidator): 잘못된 에러 메시지 수정

yeongipark committed:
refactor(index.html): 시멘틱 태그로 변경

- 의미 없는 div태그 대신 시멘틱 태그를 사용

yeongipark committed:
style(index.css): 시멘틱 태그에 대한 스타일 추가

yeongipark committed:
refactor(index): class 이름 컨벤션 일치 시킴

yeongipark committed:
refactor: css 파일 분리

- 기존 index.css에만 존재하던 css를 분리

yeongipark committed:
feat: 스파게티 코드로 기능 동작

yeongipark committed:
refactor(dom): 접근해야 하는 노드 변수 저장

yeongipark committed:
refactor(buyHandler): 구매 로직 분리

yeongipark committed:
refactor(webUtil): 웹 유틸 분리

yeongipark committed:
refactor(resultHandler): 결과 확인 로직 분리

yeongipark committed:
refactor(restartHandler): 재시작 로직 분리

yeongipark committed:
refactor(modalHandler): modal 로직 분리

yeongipark committed:
refactor(index.html): html에 js 파일 연결

yeongipark committed:
chore: handler 파일들 위치 변경

yeongipark committed:
chore: 주석 제거

yeongipark committed:
refactor(step2-index): 해당 파일에 로직 제거

- 스파게티 코드로 작성된 코드를 모듈화하여 해당 파일의 코드 삭제

yeongipark committed:
chore: lottoMachine 변수명 변경

- "$"는 노드를 표현하는 변수에만 붙이는 걸로 수정

yeongipark committed:
chore(webUtil): 노드를 담는 변수에 "$" 표시 추가

yeongipark committed:
chore(PurchasePriceValidator) : 금액 최대값 수정

- 실제 로또 구매 최대값인 100,000으로 최대값 수정

yeongipark committed:
refactor: 클래스 이름 일관성 있게 변경

yeongipark committed:
refactor(buyHandler): form태그를 사용

- 구입 기능에 웹 접근성 향상을 위해 form 태그 사용

yeongipark committed:
refactor: 결과 출력 버튼도 form 태그로 변경

yeongipark committed:
refactor(index.html): 필요없는 값들 삭제

yeongipark committed:
Merge branch 'step1' into step2

yeongipark committed:
refactor: modal 변수에도 "$" 추가

yeongipark committed:
feat(modalHandler): modal 배경 클릭하면 닫힘 기능 추가

yeongipark committed:
chore(index.html): form 태그 위치 수정

yeongipark committed:
refactor(buyHandler): 로또 출력하는 로직 함수로 분리

- 함수로 분리하여 전체 흐름을 이해하기 쉽게 만듬

yeongipark committed:
refactor: lottoMachine 변수 새로운 파일로 이동

- dom.js에서는 html 노드만 저장하도록 하기 위해 lottoMachine 변수는 새로운 파일로 분리

yeongipark committed:
chore(parser): parser 파일명 변경

yeongipark committed:
chore: modal 클래스 일관성 있게 변경

- modal 이라는 접두사가 앞에 오도록 수정

yeongipark committed:
refactor: index.css에 존재한 스타일 파일로 분리

yeongipark committed:
refactor(buyHandler): createLottos 메서드 성능 향상을 위한 로직 변경

yeongipark committed:
refactor(resultHandler): displayResult 함수 로직 개선

- 반복해서 생성되는 counts 변수를 한번만 생성되도록 수정

LAH1203 commented on src/webView/dom.js:
```diff
@@ -0,0 +1,22 @@
+const domRefs = {
+  $priceForm: document.querySelector(".paper_price_form"),
+  $buyButton: document.querySelector(".paper_buy_button"),
+  $inputPrice: document.querySelector(".paper_price_input"),
+  $lottoInfoWrap: document.querySelector(".paper_lotto_information_wrap"),
+  $lottoInfo: document.querySelector(".paper_lotto_information"),
+  $lottoCount: document.querySelector(".paper_lotto_count"),
+  $winningForm: document.querySelector(".paper_winning_form"),
+  $modal: document.querySelector(".modal_result"),
+  $paper_winning_number_inputs: document.querySelectorAll(
+    ".paper_winning_number_input"
+  ),
+  $paper_bonus_number_input: document.querySelector(
+    ".paper_bonus_number_input"
+  ),
+  $modalCloseButton: document.querySelector(".modal_close_button"),
+  $modalRestartButton: document.querySelector(".modal_restart_button"),
+  $matchCounts: document.querySelectorAll(".modal_match_count"),
+  $winningRate: document.querySelector(".modal_winning_rate"),
+};
```
이렇게 요소 자체를 저장해서 사용해주신 게 신기했어요~ 단순히 선택자만 상수화하는 게 아니라 요소 객체를 상수화해주신 이유가 있을까요?

지금처럼 별도의 돔에 그려주는 요소에 대한 조작이 없을 경우에는 만들어주신 것처럼 domRefs로 엘리먼트를 저장하는 상수를 두는 방법도 좋은 것 같아요!
다만 만약 탭이 생긴다면? 탭에 따라 그려지는 화면이 달라져야 한다면? 페이지 내 동적으로 추가되거나 변경되는 요소가 있다면 초기 DOM 선택 시점에 존재하지 않는 요소는 참조되지 않으므로, 이 경우 적절한 업데이트 로직이나 lazy query가 필요할 수 있겠죠?!
또, 컴포넌트 단위로 나눈다고 했을 때도 그 컴포넌트에 한정된 요소는 오히려 컴포넌트 내부에서 관리하는 편이 훨씬 코드가 명확해질 것 같다고 생각해서, 한번 이렇게 만드신 배경을 여쭤보고 싶었어요~!

yeongipark replied:
>탭에 따라 그려지는 화면이 달라져야 한다면? 페이지 내 동적으로 추가되거나 변경되는 요소가 있다면 초기 DOM 선택 시점에 존재하지 않는 요소는 참조되지 않으므로, 이 경우 적절한 업데이트 로직이나 lazy query가 필요할 수 있겠죠?!
또, 컴포넌트 단위로 나눈다고 했을 때도 그 컴포넌트에 한정된 요소는 오히려 컴포넌트 내부에서 관리하는 편이 훨씬 코드가 명확해질 것 같다고 생각해서,

저도 위의 하리의 말에 정말 정말 동의해요!! 

요소를 한 곳에서 관리하면 전역적으로 사용되는 요소의 일관성을 유지할 수 있어 유지보수성과 가독성이 높아질 수 있다고 생각하여 요소 객체를 상수화 하였습니다! 하지만 하리의 리뷰를 보고 나니, 특정 컴포넌트에서만 사용되는 요소는 해당 컴포넌트 내부에서 관리하는 것이 응집도를 높이고 유지보수성을 향상시키는 데 더 적절하다고 생각이 들어요!!!

yeongipark replied:
수정 커밋 : [refactor: 특정 파일에서만 사용되는 요소 변수 사용되는 파일에서 선언되도록 이동](https://github.com/woowacourse/javascript-lotto/pull/390/commits/be8b63aea63fa6499203ac24cf1b78c5ba32d75a)

LAH1203 commented on index.html:
```diff
@@ -3,14 +3,142 @@
   <head>
     <meta charset="UTF-8" />
     <title>🎱 행운의 로또</title>
+    <link rel="stylesheet" href="./index.css" />
```
사용하지 않는 파일은 지우면 어떨까요?

yeongipark replied:
네!!

yeongipark replied:
수정 커밋 : [chore(index.html): 사용하지 않는 파일들 삭제](https://github.com/woowacourse/javascript-lotto/pull/390/commits/7bdb1011063b51987093458db840f35a7498af01)

LAH1203 commented on index.html:
```diff
@@ -3,14 +3,142 @@
+          </tbody>
+        </table>
+        <p class="modal_winning_rate"></p>
+        <button class="modal_restart_button">다시 시작하기</button>
```
버튼에 type을 명시하면 어떨까요? 지금은 모달 내에서 폼을 사용하는 것 같지 않아서요!

yeongipark replied:
버튼에 type을 명시하지 않으면 submit으로 기본 동작된다는 것을 처음 알았어요...!! 감사합니다! :) :) :)

yeongipark replied:
수정 커밋 : [feat(index.html): form에 속하지 않는 button 태그에 type 설정](https://github.com/woowacourse/javascript-lotto/pull/390/commits/7b360c738c31827d314b3ac50223ffc91bd159b1)

LAH1203 commented on index.html:
```diff
@@ -3,14 +3,142 @@
+      <main class="paper">
+        <h1 class="paper_title">🎱 내 번호 당첨 확인 🎱</h1>
+        <section class="paper_section">
+          <p>구입할 금액을 입력해주세요.</p>
```
단순히 p 태그 말고 input에 대한 label로 적어보면 어떨까요? label을 클릭했을 때 input에 포커스되도록 연결해도 좋을 것 같아요~

yeongipark replied:
조아요!!

yeongipark replied:
수정 커밋 : [feat(index.html): p태그 대신에 label 태그 사용](https://github.com/woowacourse/javascript-lotto/pull/390/commits/a300d758e0078dc53f75d849cf90750615b30e9b)

LAH1203 commented on index.html:
```diff
@@ -3,14 +3,142 @@
+          <p>구입할 금액을 입력해주세요.</p>
+          <form class="paper_price_form">
+            <div class="paper_input_price_wrap">
+              <input class="paper_price_input" type="text" placeholder="금액" />
```
숫자를 입력하는 input에서는 다른 type을 사용하는 게 좋을 것 같습니다! 아래 당첨 번호 쪽도요!

추가로, required 속성도 활용해보면 어떨까요?

yeongipark replied:
> 숫자를 입력하는 input에서는 다른 type을 사용하는 게 좋을 것 같습니다! 아래 당첨 번호 쪽도요!

네!!

> 추가로, required 속성도 활용해보면 어떨까요?

form 태그 내부의 input태그에 required를 설정하면 해당 input이 비었을 때 폼이 제출되는 것을 방지할 수 있군요!!!! 감사합니당 :) :-)

yeongipark replied:
수정 커밋 : [feat(index.html): input 속성 추가](https://github.com/woowacourse/javascript-lotto/pull/390/commits/45ace2fe60036801c6cb2a259d25e9e60b8f8297), [style(init.css): input type=number의 화살표 제거](https://github.com/woowacourse/javascript-lotto/pull/390/commits/d247bd99a224604d3d39e8f0203b409a4c159baf)

LAH1203 replied:
화살표 제거 스타일도 적용해주셨군요 ㅎㅎㅎ 좋아요~!

LAH1203 commented on src/style/app.css:
```diff
@@ -0,0 +1,6 @@
+#app {
+  height: 100vh;
```
vh와 dvh의 차이는 무엇일까요?

yeongipark replied:
찾아보기 전
- vh: 현재 보고 있는 화면의 높이를 기준으로 100%
- dvh: vh와 동일하지만, 모바일 디바이스에서 화면 높이 변동에 대응

찾아본 후
- vh: 기기의 전체 뷰포트 높이를 기준으로 계산되는 단위.
	- 모바일 브라우저 UI(주소창, 네비게이션 바 등)를 포함한 전체 화면 높이를 기준으로 설정됨.
- dvh: 모바일 브라우저에서 UI(주소창, 네비게이션 바 등)의 변동을 반영하는 동적 단위.
	- 주소창이 사라지거나 다시 나타날 때 자동으로 높이가 조정됨.

웹에서는 vh와 dvh가 동일하게 동작하지만, 모바일 환경에서도 자연스럽게 스타일을 유지하려면 dvh를 사용하는 것이 더 적절하다고 생각해요!

계속해서 찾아볼 거리를 주셔서 감사합니다~~! 😊

yeongipark replied:
수정 커밋 : [style(app.css): vh->dvh로 변경](https://github.com/woowacourse/javascript-lotto/pull/390/commits/30d12da1861f5296033028dc451931ddf89f16e7)

LAH1203 commented on src/style/footer.css:
```diff
@@ -0,0 +1,9 @@
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  border-top: 1px solid #4e5ba633;
```
여기 border 색상은 css variable로 정의해주지 않으신 이유가 있나요? 뭔가 다른 건 다 정의되어 있는데 여기는 적용해주지 않으셔서 살짝 이질감이 있는 것 같아서요!

yeongipark replied:
해당 색상은 한 번만 사용되기 때문에, 굳이 CSS 변수로 설정할 필요가 없다고 판단하여 따로 정의하지 않았습니다.
또한, CSS 변수로 저장한 색상은 구현 피그마에서 제공된 색상들만 포함하도록 했습니다.

일회성으로 한 번만 사용되는 색상도 CSS 변수로 설정하는 것이 더 좋은 방법일까요?
하리의 의견이 궁금합니다! 😊

LAH1203 replied:
음 이건 회사 / 팀마다 의견이 다를 것 같기는 한데, 저 같은 경우는 설정해두는 편입니다~!
다만 응집의 목적이 큰 것 같아요. 나중에 색상 변경이 있더라도 정의해둔 팔레트 내에서 수정이 가능하다는 점이 좋더라고요! 그리고 보통 디자이너 분들도 팔레트 내에서 대부분의 색상을 충당하시기도 하고요..!
그래서 좀 다른 얘기지만 저는 팔레트를 중요하게 생각하고 있고, 언제든 다른 곳에서도 사용할 수 있도록 색상은 최대한 보편적으로 네이밍하고 있습니당 (ex : button-active(x), blue(o) )

LAH1203 commented on src/style/init.css:
```diff
@@ -0,0 +1,72 @@
+  font-weight: 400;
+}
+
+button .caption {
```
init에서 특정 클래스명에 대한 스타일을 정의하는 건 적절할까요?

yeongipark replied:
아뇨..!

특정 컴포넌트나 페이지에 한정된 클래스를 정의하는 것은 해당 컴포넌트 파일에 정의하는게 좋다고 생각해요!!

하지만 해당 스타일은 실제로 사용되지 않으니 제거하는게 좋을 거 같습니다!

yeongipark replied:
수정 커밋 : [chore(init.css): 사용되지 않는 스타일 삭제](https://github.com/woowacourse/javascript-lotto/pull/390/commits/c1eb3a212d05ae3ebc3b481134988098396a7b6c)

LAH1203 commented on src/style/modal.css:
```diff
@@ -0,0 +1,62 @@
+}
+
+.modal_statistics {
+  width: 319px;
+  height: 247px;
+  margin-top: 33px;
```
팁 : 픽셀은 홀수 단위보다는 짝수 단위로 표현하면 좋습니다~! 예전에는 4px 단위로 맞추라는 것도 있었지만... 지금은 그렇게까지 빡빡하게 할 필요는 없을 것 같고 대신 짝수 단위로만 맞춰도 괜찮더라고요!

브라우저는 레이아웃 계산 시 소수점 이하의 값을 반올림하는 경우가 많은데, 홀수 픽셀을 사용하면 이 과정에서 경계가 불분명해지거나 오차가 생길 수 있어 보통 짝수 픽셀로 많이 디자인하고 사용하는 것 같아요~

yeongipark replied:
오..! 처음 안 사실이네요!! 감사합니다 😊

yeongipark replied:
수정 커밋 : [style: 홀수 px -> 짝수 px로 변경](https://github.com/woowacourse/javascript-lotto/pull/390/commits/1c24f1d11f96d080802f305dfffa18abf642a7f8)

LAH1203 commented on src/style/paper.css:
```diff
@@ -0,0 +1,97 @@
+  background-color: var(--grayscale1);
+  border: 1px solid rgba(0, 0, 0, 0.12);
+  margin: 0 auto;
+  padding: 40px 16px 32px 16px;
```
이건 `padding: 40px 16px 32px;` 로도 축약할 수 있을 것 같아요~!

yeongipark replied:
오!! 3개로 하는 방법도 있었군요! 감사합니당

yeongipark replied:
해당 수정 커밋은 styleLint 적용하면서 자동으로 바뀌었네요 ㅜㅜ

yeongipark replied:
수정 커밋 : [chore(style): CSS 컨벤션 적용](https://github.com/woowacourse/javascript-lotto/pull/390/commits/746132839de9d70ffb9d45281f6b1d3a9859e877)

LAH1203 replied:
ㅎㅎ 그렇군요 확인했습니다~!
나중에 `padding: 40px 16px 32px 16px;`을 `padding: 40px 16px 32px;`로 축약하면 뭐가 좋을지! 한번 궁금하실 때 찾아보셔도 좋을 것 같아요~! 크게 중요하지는 않아서, 궁금할 때만 찾아보셔요 ㅋㅋㅋ
(비슷한 측면에서 `0px` 대신 `0`을 쓰라는 것도 있어영)

LAH1203 commented on src/style/paper.css:
```diff
@@ -0,0 +1,97 @@
+  input {
+    width: 310px;
+    height: 36px;
+    padding: 8px 8px;
```
이것도 padding이 모든 방향에서 동일하다면 한번만 적어도 될 것 같습니다 :)

yeongipark replied:
넹!~~~

yeongipark replied:
수정 커밋 : [chore(style): CSS 컨벤션 적용](https://github.com/woowacourse/javascript-lotto/pull/390/commits/746132839de9d70ffb9d45281f6b1d3a9859e877)

LAH1203 commented on src/style/paper.css:
```diff
@@ -0,0 +1,97 @@
+  margin-bottom: 8px;
+}
+
+.paper_winning_input_wrap {
+  display: flex;
+  justify-content: space-between;
+
+  input {
+    width: 34px;
+    height: 36px;
+    padding: 8px;
+  }
+}
```
보통 이렇게 스타일링 하는 건 돔에 크게 의존하고 있는 스타일링이어서 추천하지는 않는 방식이에요!
필수는 아니지만, 명시적으로 input에 class를 부여하고 스타일링하는 방식을 추천드립니당

yeongipark replied:
네!!

그러면 혹시 아래와 같은 경우도 DOM 구조에 크게 의존하는 스타일이므로, 지양하는 것이 좋은 방식인가요?? 🤔

```css
.paper_winning_input_wrap > .작명한input{
    width: 34px;
    height: 36px;
    padding: 8px;
}
```

yeongipark replied:
수정 커밋 : [refactor(paper.css): 중첩 CSS를 개별 CSS 코드로 분리](https://github.com/woowacourse/javascript-lotto/pull/390/commits/95f9885405ef9b6148756e3c481a36fdb635d9eb)

LAH1203 replied:
말씀해주신 경우는 괜찮습니다 😊 이유는 `.작명한input` 같이 커스텀 클래스로 정의해주신 방식 때문인데요~!
이전에는 기본 태그에 스타일이 의존하고 있었기 때문에 만약 돔에 다른 input이 추가되면 그에 동일한 스타일이 적용될 위험이 있었어요. 이 부분에서 돔에 크게 의존하고 있다고 말씀 드렸던 거였고요. 하지만 커스텀 클래스를 사용하면 의도한 범위 내에서만 동작하도록 명시할 수 있어 돔 구조에 변화가 생기더라도 크게 영향을 받지 않겠죠?! 지금보다 의존성이 명확하고 제어가 가능하다는 점에서 훨씬 좋다고 생각합니당

yeongipark replied:
감사합니당:):)

LAH1203 commented on src/util/webUtil.js:
```diff
@@ -0,0 +1,27 @@
+export function disabled($tags, bool) {
+  $tags.forEach((tag) => {
+    tag.disabled = bool;
+  });
+}
```
함수명을 좀 더 명시적으로 바꾸면 어떨까요?
이 함수는 태그 목록을 받아서 disabled 상태를 일괄 변경해주는 역할을 하는 것 같은데, 단순히 `disabled` 로는 함수의 로직을 보지 않고서는 역할을 파악하기 힘들 것 같아요~

yeongipark replied:
네!!

yeongipark replied:
수정 커밋 : [refactor: disabled -> setTagsDisabled로 함수명 변경](https://github.com/woowacourse/javascript-lotto/pull/390/commits/236e3f0decbaa96e93303006d193ae83975f4468)

LAH1203 commented on src/webView/buyHandler.js:
```diff
@@ -0,0 +1,57 @@
+  domRefs.$lottoCount.textContent = `총 ${lottosNumber.length}개 구매했습니다.`;
+
+  disabled([domRefs.$inputPrice, domRefs.$buyButton], true);
+  focusFirstNode(domRefs.$paper_winning_number_inputs);
```
자동으로 포커스를 이동하도록 해주셔서 실행할 때 편하더라고요 😆

LAH1203 commented on src/webView/buyHandler.js:
```diff
@@ -0,0 +1,57 @@
+  }
+});
+
+function displayLottoInfo(lottoMachine) {
```
함수명은 로또 정보를 보여주는 역할을 하는 것 같은데, 내부를 보면 로또 정보를 보여주고, 가격 입력 폼을 disabled하고, 당첨 번호 인풋에 포커스를 하는 여러 가지 동작을 하고 있어요~!
여기에서는 함수를 분리해보면 어떨까요? 가격 입력 폼을 disabled하고인풋에 포커스를 하는 과정은 display와는 분리되어야 할 후처리 동작 같아요!

yeongipark replied:
동의해요!! 분리해보게씁니다!

yeongipark replied:
수정 커밋 : [refactor(buyHandler.js): displayLottoInfo함수 내부 불필요한 동작 제거](https://github.com/woowacourse/javascript-lotto/pull/390/commits/349ad4e97decb01f428a71c61733e16cff98b0af)

LAH1203 commented on src/webView/buyHandler.js:
```diff
@@ -0,0 +1,57 @@
+    const lottoNumbersDiv = createElement({
+      tag: "div",
+      className: "lotto_number",
+      text: numbers.sort((a, b) => a - b).join(", "),
```
텍스트를 포맷해주는 로직은 별도의 유틸 함수로 분리해볼 수 있을 것 같아요~

yeongipark replied:
네!!

yeongipark replied:
수정 커밋 : [refactor(buyHandler.js): 숫자 배열 정렬후 문자열 반환하는 로직 유틸 함수로 분리](https://github.com/woowacourse/javascript-lotto/pull/390/commits/490fc329ce82a21d0590d0b6f07d8e558dc5df09)

LAH1203 commented on src/webView/buyHandler.js:
```diff
@@ -0,0 +1,57 @@
+function createLottos(lottosNumber) {
+  const fragment = document.createDocumentFragment();
+
+  const lottoDivs = lottosNumber.map((numbers) => {
+    const lottoDiv = createElement({ tag: "div", className: "lotto" });
+    const ticketIcon = createElement({
+      tag: "div",
+      className: "ticket_icon",
+      text: "🎟️",
+    });
+    const lottoNumbersDiv = createElement({
+      tag: "div",
+      className: "lotto_number",
+      text: numbers.sort((a, b) => a - b).join(", "),
+    });
+
+    lottoDiv.appendChild(ticketIcon);
+    lottoDiv.appendChild(lottoNumbersDiv);
+    return lottoDiv;
+  });
+
+  lottoDivs.forEach((lottoDiv) => fragment.appendChild(lottoDiv));
```
여기에서는 반복문을 두번 돌 필요 없이 한번으로 처리할 수 있을 것 같네요~!

yeongipark replied:
네!

yeongipark replied:
수정 커밋 : [refactor(buyHandler): createLottos 함수에서 불필요한 반복문 삭제](https://github.com/woowacourse/javascript-lotto/pull/390/commits/8d2d11340f6acbc42eb455a6339ee31e5ddd53fc)

LAH1203 commented on src/webView/restartHandler.js:
```diff
@@ -0,0 +1,15 @@
+import domRefs from "./dom.js";
+import { disabled, initNode, initNodes } from "../util/webUtil.js";
+
+domRefs.$modalRestartButton.addEventListener("click", () => {
```
보면 모든 이벤트 핸들러가 add만 되고 remove는 이뤄지지 않더라고요~! 지금의 구조에서는 문제가 없지만, 연습하는 측면에서 추가해보면 어떨까요?
만약 remove를 추가한다면 어떤 시점에서 이벤트 핸들러를 제거하면 좋을까요?

yeongipark replied:
메모리 누수를 방지하고, 효율적으로 메모리를 관리하는 연습을 하면 좋을 것 같아요!
제가 생각한 방식은 아래와 같습니다!

1️⃣ 구매 버튼 (buy 버튼)
- 구매 버튼을 누르면 구매 버튼 이벤트 핸들러를 제거
- 다시 시작 버튼을 누르면 구매 버튼 이벤트 핸들러를 다시 추가

2️⃣ 결과 확인 버튼 
- 결과 확인 버튼을 누르면 결과 확인 버튼 이벤트 핸들러를 제거
- 구매 버튼 or 모달창을 닫으면 결과 확인 버튼 이벤트 핸들러를 추가

3️⃣ 모달의 다시 시작 버튼 & 닫기 버튼
- 다시 시작 버튼 누르면 모달 이벤트 핸들러를 제거
- 결과 확인 버튼을 누르면 모달 이벤트 핸들러 추가

yeongipark replied:
수정 커밋 : [feat: 이벤트 remove 기능 추가](https://github.com/woowacourse/javascript-lotto/pull/390/commits/d167029ff2dd20d1a0b4a58d8451599c42466d87)

LAH1203 replied:
크 모달을 기준으로 적절한 시점에 적용해주셨네요...! 👍

LAH1203 reviewed (changes requested):
안녕하세요 우디~! 2단계 정말 열심히 해주셨네요!! 보면서 많이 감탄했습니당

긴 말 없이, 우디가 PR에 많은 질문을 주셔서 그 부분 답변부터 해보겠습니다 ㅎ.ㅎ

> 파일을 어느 폴더에 저장할지 고민하다가, **웹뷰와 관련된 코드들이므로 webView 폴더에 저장**했습니다. 하지만 지금 다시 생각해보니, **핸들러 파일들이 주로 이벤트 처리를 담당하므로 eventHandler라는 폴더명이 더 적절할 수도 있겠다는 생각이 드네요**

동의합니다! 엄밀히 따지면 1단계에서 우디가 만들어주셨던 controller의 역할을 이벤트 핸들러에서 담당하고 있다고 볼 수 있을 것 같아요.

> 만약 각 파일에서 **import하는 변수가 2~3개 정도였다면 개별적으로 export하는 방식이 더 적절**했을 것 같습니다. 하지만 **일부 파일에서 import 문이 너무 많아지면서 코드의 가독성이 저하된다고 판단**하여, 결국 객체로 묶어서 export하는 방식을 선택했습니다.
> 이에 대해 **하리의 의견이 궁금합니다!** 😊

저는 import문이 많아진다고 가독성이 떨어진다고 생각하지는 않아요! 파일 내에 있긴 하지만, 로직을 파악할 때 굳이 import문까지 보지는 않거든요. 그래도 지금처럼 객체로 묶어주신 방식은 응집도 측면에서 좋다고 생각합니다~!

> • **함수를 찾아가는 시간** vs **로직을 바로 이해하는 시간**이 비슷할 수 있음
> • **재사용되지 않는 함수라면 분리의 이점이 크지 않음**
> 이러한 이유로 결국 **별도로 분리하지 않고 그대로 두기로 결정**했는데, 이에 대해 **하리의 의견이 궁금합니다!** 😊

오... 저는 재사용되지 않는 함수여도 분리의 이점이 크다고 생각합니당.
우리가 함수를 분리하는 목적이 단순히 "재사용을 위해서"만일까요? 저는 함수를 분리함으로 인해 로직을 이해하는 데 큰 도움을 준다고 생각해요.
만약 특정 로직을 이해하는 데 함수를 찾아가야 한다면, 이는 함수를 잘못 정의한 형태일 것 같아요. 함수명이 그 함수가 하는 일을 정확히 명시하고 있다면 굳이 함수를 찾아갈 일도 없을 것 같고, 함수를 분리하지 않아도 이미 함수가 한 가지 행동을 하고 있었다면 과도한 분리이지 않을까요?

```js
Array.from(state.paper_winning_number_inputs).map(
	(winningInput) => Number(winningInput.value)
);
```

우디가 예시로 들어주신 코드를 볼게요.
위 로직은 input 배열의 값을 숫자로 변환해주는 한 가지의 역할을 이미 수행하고 있다고 생각합니다. 굳이 이걸 또 input 배열로 만드는 로직, input 배열을 순환하는 로직, input 값을 숫자로 변환하는 로직으로 나누어 각각의 함수로 분리하는 건 과도한 분리일 것 같아요!
그래서 함수를 별도로 분리해주지 않으신 건 저는 좋다고 생각해요~!

한번 함수 분리의 기준을 세워보시면 어떨까요? 가장 간단한 기준은 '이 함수가 하는 역할을 하나로 표현할 수 있는가'일 것 같아요. 그 역할을 어떻게 문장으로 표현할 수 있을지 적어봅시당. 만약 단순히 하나의 동사가 아닌 여러 개의 동사가 내포되어야 한다면, 여러 가지의 역할을 하고 있는 함수라고 말할 수 있을 것 같아요.

> 또한, **DOM 조작은 큰 비용이 소모되는 작업이므로 최소화하는 것이 중요하다고 생각하여, 정말 필요한 부분만 JavaScript로 동적으로 렌더링**하기로 결정했습니다.

아주 좋은 기준으로 작성해주셨네요~!
관련해서 같이 공부해보시면 좋을 것 같은 포인트를 말씀드리자면, 브라우저가 html, css, js 코드를 어느 시점에서 실행하고 가져오는지도 비교해보시면 좋을 것 같아요! (키워드 : `브라우저 렌더링 과정`)
이미 알고 계시다면 훌륭합니다 😎

> 이때까지 css도 컨벤션을 지켜서 개발한다는 사실을 모르고 이번에 알았습니다. 그래서 저만의 css 컨벤션 규칙을 만들어서 이번에 적용해보았습니다! 혹시 컨벤션이 제대로 지켜지지 않은 부분은 얘기해 주세요!

오 css도 규칙을 세워서 지키려고 하셨군요! 너무 좋아요!!!
그러면 [stylelint](https://stylelint.io/) 도 활용해보시는 건 어떨까요? 만들어주신 규칙을 잘 지키고 있는지 도움을 많이 줄 수 있을 것 같아요 ㅎㅎ

> 각 방법의 장단점을 비교해본 결과, 1번 방식(link 태그 사용)은 CSS 파일을 비동기적으로 병렬 다운로드하여 페이지 로딩 속도가 더 빠르다는 장점이 있었습니다.
> 반면, 2번 방식([@import](https://github.com/import) 사용)은 CSS 파일이 순차적으로 로드되기 때문에 성능이 상대적으로 떨어집니다.
> 이러한 점을 고려하여, 최적의 성능을 위해 1번 방식( 태그 사용)을 선택했습니다.

우디 2단계를 정말 열심히 공부하시면서 구현해주셨군요! 짱짱
한 가지 여쭤보고 싶어요!
1번 방식을 사용했을 때 페이지 로딩 속도가 빨라지는 건 단순히 CSS 파일을 비동기적으로 병렬 다운로드하기 때문만일까요?! (키워드 : `HTML parser 동작`)

> 이전의 createLottos 함수에서는 로또를 생성할 때마다 lottoInfo(로또를 담는 컨테이너)에 appendChild를 호출하여, 로또 개수만큼 appendChild가 실행되는 문제가 있었습니다.
> 하지만 appendChild는 DOM을 직접 조작하는 작업이므로, 호출 횟수를 최소화하는 것이 성능 개선에 중요하다고 판단했습니다.
> 따라서, 로또를 생성할 때마다 즉시 lottoInfo에 추가하는 것이 아니라, DocumentFragment에 먼저 추가한 후, 모든 로또가 생성된 후에 단 한 번만 appendChild를 실행하도록 변경하여 성능을 개선했습니다.

너무 훌륭한데요...
말씀해주신대로 appendChild는 최소화할수록 DOM에 직접 접근하는 경우가 줄어들기 때문에 효과적인 최적화가 되죠!
여기서도 한가지 여쭤보고 싶어요!
성능을 개선하기 위해 이렇게 해주셨다고 했는데, 왜 appendChild를 사용해 매번 DOM을 조작하는 게 성능에 좋지 않을까요? (키워드 : `리플로우 / 리페인트`)

> 저는 평소에 id보다 class를 더 많이 사용합니다!
> 그 이유는 **처음 리액트와 HTML을 배울 때 인강에서 id를 사용하지 않고 class를 주로 사용했기 때문**입니다.
> 보통 **id는 고유한 값을 나타낼 때, class는 재사용이 가능할 때 사용한다고 알고 있는데**, 실제로 **현업에서도 이 원칙을 그대로 적용하는지 궁금합니다.**
> 또한, 제 코드에서는 **한 번만 사용되는 class가 많습니다.**
> 이 경우 **모두 id로 바꾸는 것이 좋은지**에 대한 하리의 의견이 궁금합니다!

선택자를 부여하는 목적에 따라 다르다고 생각합니다~! 
말씀해주신 것처럼 id는 고유한 식별자를 부여할 때, class는 재사용이 필요할 때 주로 사용하는 편이기는 해요. 만약 두 가지의 목적이 모두 필요하다면 굳이 둘 중 하나를 고를 필요 없이 둘 다 사용하면 되겠죠?? 하지만 그 외에도 스타일링을 위해서는 한번만 사용하더라도 class를 사용하는 편이 더 유연하고 유지보수가 용이하다고 생각합니당
이렇게 말하면 왜 class를 사용하는 게 더 유연하고 유지보수가 용이한 것인지 궁금하실 것 같아요! 혹시 css에서 선택자의 우선순위에 대해 알고 계신가요? id로 정의된 스타일과 class로 정의된 스타일에서 충돌되는 일부 스타일 요소가 있을 때, 어떻게 오버라이딩될까요?

> 저는 보통 **부모 요소의 이름을 앞에 작성한 후, _(언더스코어)로 자식 요소를 구분하는 방식**을 사용합니다.
> 예를 들어, paper_input처럼 네이밍을 합니다.
> 이렇게 하는 이유는 **클래스명이 의도치 않게 겹쳐 스타일이 깨지는 것을 방지하기 위해서입니다.**
> 혹시 **하리는 다른 네이밍 방식을 사용한다면 어떻게 하는지** 궁금하고, **제 방식에 대한 의견도 듣고 싶습니다!**

저는 개인적으로 [BEM 방식](https://getbem.com/naming/) 을 선호하는데, 우디가 적용해주신 방식은 BEM의 변형인 것 같아요! 지금처럼 module css를 사용하지 않고 전역 css로 관리할 때는 만들어주신 것처럼 스타일링의 범위를 명시해주면 좋다고 생각합니당

> CSS를 작성하다 보면 container와 wrap을 사용하는 경우가 많습니다.
> 저는 **최상위 요소를 container로 설정하고, 그 내부를 묶는 용도로 wrap을 사용**합니다.
> 실제로 **이렇게 사용하는 것이 맞는지**, 혹은 다른 기준이 있는지 궁금합니다!

저도 우테코할 때 이걸로 많이 고민했던 것 같아요! 결론은 "정답은 없습니다"! ㅋㅋㅋㅋㅋ
저도 보통 최상위를 컨테이너, 그 내부에서 래퍼, 섹션 등으로 나눴었어요. 이 부분은 규칙을 세워서 일관성만 잘 지켜주시면 좋을 것 같아요~

> 이번에 Web으로 구현하면서 컨트롤러가 없어지고, addEventListener를 통해 모든 사용자 입력과 출력을 관리하게 되었습니다. 
> 그렇다 보니, 1단계에서와는 달리 step2-index.js 파일을 활용할 방법이 없었습니다.
> 이전처럼 해당 파일에서 실행을 명시적으로 호출할 필요도 없어, 이번에는 step2-index.js에 아무 내용도 작성하지 않았는데, 하리님께서는 이 파일에 어떤 내용을 넣는 것이 좋을지, 또는 구조를 어떻게 변경하면 좋을지에 대한 의견이 궁금합니다!

음 개별 js 파일을 script로 각각 index.html에서 주입해주셔서 그런 것 같아요~! 아마 원래는 각 js 파일을 `step2-index.js`에서 호출하라는 의도가 있지 않았을까요..?!
지금은 각각의 js 파일을 별개의 script 태그로 주입해주셨는데, 이렇게 했을 때 어떤 측면에서 다를까요?

> 앞서 고민한 내용에서도 언급했듯이, 저는 최대한 HTML에서 직접 구조를 작성하고, 정말 필요한 부분만 JavaScript로 동적으로 렌더링하도록 했습니다.
> 하지만 많은 크루들이 모든 HTML을 JavaScript로 렌더링하는 방식을 선택하더라고요.  
> 처음에는 제 방식이 더 적절하다고 생각하여 그대로 진행했지만, 점점 주변 크루들이 JavaScript로 렌더링하는 것을 보면서 “내가 틀린 걸까?” 하는 생각이 들었습니다.
> 그런데 아무리 고민해도 JavaScript로 모든 HTML을 렌더링했을 때의 장점이 명확하게 와닿지 않습니다.  
> HTML 파일이 깔끔해진다는 점 정도는 이해가 가지만, 그 외에 JavaScript를 통한 렌더링이 가지는 구체적인 이점이 무엇인지 궁금합니다.  
> 또한 해당 방식이 정말 html을 직접 작성하는 것보다 좋은 건지 궁금해요!!
> 하리는 이 부분에 대해 어떻게 생각하나요 !?!?

각자의 장단점이 뚜렷해서 정답은 없다고 생각합니다~!
우디가 작성해주신대로 정적인 부분은 HTML로 작성하면 초기 로딩 속도와 유지보수 측면에서 좋을 것 같아요! 반대로 모든 HTML을 자바스크립트로 렌더링하는 방식은 컴포넌트 기반 개발에서 좋을 것 같고요. 하지만 이 경우에는 자바스크립트 파일의 실행이 끝날 때까지는 화면을 볼 수 없겠죠?!
그래서 저는 우디가 작성해주신 방식도 넘넘 좋았습니다~ 지금 같은 구조에서는 적절하게 사용해주셨다고 생각해요!

---

그 외에 몇 가지 부분 코멘트를 남겼는데, 궁금하신 부분은 1단계 때처럼 지속적으로 질문 주시면 시간될 때마다 답변 드리겠습니다~! 😆

yeongipark committed:
refactor(buyHandler): createLottos 함수에서 불필요한 반복문 삭제

yeongipark committed:
chore: CSS 스타일 컨벤션 유지를 위해 styleLint 설정

yeongipark committed:
chore(style): CSS 컨벤션 적용

yeongipark committed:
chore(eventHandler): eventHandler 파일들 폴더 이동

yeongipark committed:
chore(index.html): 사용하지 않는 파일들 삭제

yeongipark committed:
feat(index.html): form에 속하지 않는 button 태그에 type 설정

yeongipark committed:
feat(index.html): p태그 대신에 label 태그 사용

yeongipark committed:
feat(index.html): input 속성 추가

- required 속성을 추가해 값이 비었을 때 form 제출을 막음

- type="number"를 적용해 숫자가 아닌 값이 입력되는 것을 막음

- max와 min 값을 넣어 유효한 값만 입력할 수 있도록 제한

yeongipark committed:
style(init.css): input type=number의 화살표 제거

yeongipark committed:
style(app.css): vh->dvh로 변경

yeongipark committed:
chore(init.css): 사용되지 않는 스타일 삭제

yeongipark committed:
style: 홀수 px -> 짝수 px로 변경

yeongipark committed:
refactor: disabled -> setTagsDisabled로 함수명 변경

yeongipark committed:
refactor(buyHandler.js): displayLottoInfo함수 내부 불필요한 동작 제거

yeongipark committed:
refactor(buyHandler.js): 숫자 배열 정렬후 문자열 반환하는 로직 유틸 함수로 분리

yeongipark committed:
feat: 이벤트 remove 기능 추가

- remove 기능을 추가하여 메모리 누수를 막음

yeongipark committed:
refactor(paper.css): 중첩 CSS를 개별 CSS 코드로 분리

yeongipark committed:
refactor: 특정 파일에서만 사용되는 요소 변수 사용되는 파일에서 선언되도록 이동

yeongipark committed:
refactor: js에서 CSS 코드 작성 -> 클래스 이름 추가 방식으로 변경

- display : block을 js 코드로 직접 등록했던 방식에서 클래스 이름을 추가하고 제거하는 방식으로 변경

yeongipark committed:
refactor(step2-index.js): 필요없는 함수 호출 제거

yeongipark commented:
@LAH1203 하리의 정성스러운 리뷰 덕분에 많은 것을 찾아보고 공부할 수 있었습니다! 특히, 브라우저 렌더링과 CSS 다운로드 방식에 대해 기존에는 얕게 알고 있었는데, 이번 기회를 통해 깊이 이해할 수 있게 되었어요. 정말 감사합니다! ☺️☺️☺️

## 주요 수정 사항
### 1️⃣ 
기존에는 JavaScript에서 직접 CSS를 조작했기 때문에, 스타일이 변경될 때 JavaScript 코드도 함께 수정해야 하는 문제가 있었습니다. 이를 방지하기 위해, CSS를 직접 변경하는 대신 클래스를 추가하고 제거하는 방식으로 수정하여 유지보수성을 높였습니다! 🚀

수정 커밋 : [refactor: js에서 CSS 코드 작성 -> 클래스 이름 추가 방식으로 변경](https://github.com/woowacourse/javascript-lotto/pull/390/commits/c049f094b75af5f84406a063f86f703682cc594a)

기존 코드
```javascript
domRefs.$lottoInfoWrap.style.display = "block";
```

수정한 코드
```javascript
domRefs.$lottoInfoWrap.classList.add("show");
```

### 2️⃣ styleLint 적용
하리의 리뷰를 통해 styleLint를 처음 적용해 보았어요!! 너무 편리하고 CSS 컨벤션을 유지하는데 너무너무 도움될 거 같아요!!

### 3️⃣ 이벤트 핸들러 remove적용
이벤트 핸들러를 정말 필요한 경우에만 등록하고, 불필요할 때는 제거하여 메모리 낭비를 최소화했습니다!

이 과정에서 여러 개의 handler 함수가 생겼고, addEventListener와 removeEventListener가 여러 곳에서 반복적으로 사용되었습니다.
이를 개선하기 위해, 특정 이벤트 핸들러를 추가 및 제거하는 함수를 만들어 export하여 여러 곳에서 재사용할 수 있도록 수정했습니다.

또한, 이벤트 핸들러를 함수로 등록할 수 있도록 로직을 수정하여, 기존에 비어 있던 step2-index.js에서 addBuyEventHandler 함수를 호출해 구매 이벤트 핸들러를 등록하도록 변경했습니다!

기존 코드
```javascript
domRefs.$priceForm.addEventListener("submit", (e) => { 
 // ... 실행 로직
}
```
수정 코드
```javascript
export function addBuyEventHandler() {
  domRefs.$priceForm.addEventListener("submit", buyHandler);
}

export function removeBuyEventHandler() {
  domRefs.$priceForm.removeEventListener("submit", buyHandler);
}
```
이로 인해, 이벤트 핸들러를 필요할 때만 효율적으로 관리할 수 있고, 코드의 재사용성이 향상되었습니다! 

---
## 궁금한 점🤔
### 1️⃣ display : none의 문제점
display: none;으로 요소를 숨기면, 개발자 도구를 통해 CSS를 조작하여 의도치 않은 이벤트가 발생할 수 있었습니다. 이런 예상치 못한 에러를 방지하기 위해 추가적인 대응 코드가 필요할까요, 아니면 display: none; 대신 해당 요소를 필요할 때 JavaScript로 동적으로 생성하는 것이 더 적절한 방식일까요? 🤔

### 2️⃣ 요소를 저장한 변수 분리
요소 변수들을 분리하면서 응집도를 높이는 것이지만, **코드를 읽는 입장에서 일관성이 떨어질 수 있다는 점이 신경 쓰였어요.**

예를 들어, **어떤 요소 변수들은 domRefs를 통해 접근하고, 특정 요소 변수들은 현재 파일에 정의되어 있다면, 코드의 일관성이 깨져 헷갈릴 수 있다고 생각했어요.**

그래서 **이왕 요소 변수를 별도의 파일에서 관리하기로 했다면, 모든 요소를 한 곳에서 관리하는 것이 오히려 코드를 이해하는 데 도움이 될 수도 있겠다는 생각이 들었어요.**

또한, 만약 특정 요소 노드를 사용하는 변수를 수정해야 하는 상황에서,
**모든 요소가 하나의 파일에 정의되어 있다면 → 해당 파일만 보면 되지만**
**여러 파일로 분리되어 있다면 → 해당 요소 변수가 어디에서 사용되고 있는지 모르고, 결국 모든 파일을 열어봐야 할 수도 있다고 생각했어요.**

이런 점들을 종합해 봤을 때, **오히려 요소 변수를 하나의 파일에서 관리하는 것이 코드의 일관성을 유지하고, 유지보수 측면에서도 더 좋을 수 있다고 생각했어요!**

이 부분에 대해 어떻게 생각하시나요? 😊

### 3️⃣ PR 메시지 고민
2단계 미션 첫 번째 리뷰 요청 PR에 고민한 내용을 너무 많이 작성해서, 리뷰어께서 읽는 데 부담을 느낄 수도 있겠다는 생각이 들었습니다.
하리께서는 어떻게 느끼셨는지 궁금합니다! 
또한,다음 PR 메시지에서는 어떤 부분을 추가하거나 줄이면 좋을지 피드백을 부탁드립니다. 😊

LAH1203 commented:
> display: none;으로 요소를 숨기면, 개발자 도구를 통해 CSS를 조작하여 의도치 않은 이벤트가 발생할 수 있었습니다. 이런 예상치 못한 에러를 방지하기 위해 추가적인 대응 코드가 필요할까요, 아니면 display: none; 대신 해당 요소를 필요할 때 JavaScript로 동적으로 생성하는 것이 더 적절한 방식일까요? 🤔

저는 일반적으로는 CSS 조작까지 대응할 필요는 없다고 생각해요! 그냥 `display: none;`만 사용해도 충분할 것 같습니다.
제가 생각하는 방어 코드를 작성하는 조건은 말씀해주신 '의도치 않은 이벤트'가 애플리케이션에서 심각한 부작용을 일으키는 경우가 될 것 같은데요~! 예를 들면 폼 submit 버튼의 disabled 해제 같은 경우가 있겠네요.
말씀해주신 동적 생성 방식을 고려하는 것도 좋지만, 이 경우에는 동적 생성 방식과 사용하는 목적이 다르기 때문에 대부분은 이벤트 핸들러 내에서의 방어 코드를 추가하는 방향으로 많이 하는 것 같습니다~! 

> 요소 변수들을 분리하면서 응집도를 높이는 것이지만, 코드를 읽는 입장에서 일관성이 떨어질 수 있다는 점이 신경 쓰였어요.
> 예를 들어, 어떤 요소 변수들은 domRefs를 통해 접근하고, 특정 요소 변수들은 현재 파일에 정의되어 있다면, 코드의 일관성이 깨져 헷갈릴 수 있다고 생각했어요.
> 그래서 이왕 요소 변수를 별도의 파일에서 관리하기로 했다면, 모든 요소를 한 곳에서 관리하는 것이 오히려 코드를 이해하는 데 도움이 될 수도 있겠다는 생각이 들었어요.
> 또한, 만약 특정 요소 노드를 사용하는 변수를 수정해야 하는 상황에서,
> 모든 요소가 하나의 파일에 정의되어 있다면 → 해당 파일만 보면 되지만
> 여러 파일로 분리되어 있다면 → 해당 요소 변수가 어디에서 사용되고 있는지 모르고, 결국 모든 파일을 열어봐야 할 수도 있다고 생각했어요.
> 이런 점들을 종합해 봤을 때, 오히려 요소 변수를 하나의 파일에서 관리하는 것이 코드의 일관성을 유지하고, 유지보수 측면에서도 더 좋을 수 있다고 생각했어요!
> 이 부분에 대해 어떻게 생각하시나요? 😊

지금과 같은 규모에서는 적절한 솔루션이라고 생각해요!
말씀해주신 방식은 두 가지 관점으로 볼 수 있을 것 같은데요.
- 중앙 집중식 관리 (domRefs)
- 모듈화 및 컴포넌트별 관리 (각 파일에서 요소 상수 관리)

지금처럼 소규모의 프로젝트에서는 중앙 집중식으로 돔 요소를 한 곳에서 관리하는 방식이 말씀해주신대로 일관성과 유지보수 측면에서 유리할 수 있을 것 같아요. 하지만 대규모 프로젝트나 컴포넌트 단위로 독립성이 중요한 경우에는 모듈화하여 각 컴포넌트 내에서 관리하는 게 캡슐화와 확장성 측면에서 유리하다고 생각합니다 ㅎㅎ

> 2단계 미션 첫 번째 리뷰 요청 PR에 고민한 내용을 너무 많이 작성해서, 리뷰어께서 읽는 데 부담을 느낄 수도 있겠다는 생각이 들었습니다.
> 하리께서는 어떻게 느끼셨는지 궁금합니다!
> 또한,다음 PR 메시지에서는 어떤 부분을 추가하거나 줄이면 좋을지 피드백을 부탁드립니다. 😊

음...! 저는 오히려 구현하신 포인트와 고민을 자세히 적어주셔서 좋았습니다 😄 오히려 그렇게 적어주셨을 때 코드를 파악하는 데 더 도움이 많이 되거든요!!! 적어주신 내용에서 크게 부담이 되었던 부분도 없었습니당 ㅎㅎ ~리뷰어에게서 최대한 많은 정보를 뽑아가세요~

LAH1203 reviewed (approved):
우디 2단계도 고생 많으셨습니다~! 🤗
고민 포인트를 자세하게 공부하고 전달해주셔서, 저도 재밌게 리뷰할 수 있었습니다 ㅎㅎ

다음 미션도 화이팅입니닷 💪💪
